Rewrite broadcast() and map() based on lift()

nalimilan opened the pull request:
Remove the custom implementation of broadcast(), and just call
the base method on the lift()ed method. This implements the
blacklist approach: methods with a custom lifting behavior like
isnull() should override lift() to get passed Nullable values
directly; if they do not return a Nullable, the result is a standard
Array rather than a NullableArray.

Performance will probably be worse than before, but at least the
semantics will be correct. We can always re-implement the custom
and faster versions later when broadcast() has stabilized in Base
and Nullable support has settled.

This is WIP since I haven't thought much about this yet, because it copies `lift` from https://github.com/JuliaLang/julia/pull/18758 without too much thinking either, ~~and because it only works on 0.6 at the moment~~.

Cf. discussions at https://github.com/JuliaStats/NullableArrays.jl/issues/144 and https://github.com/JuliaStats/NullableArrays.jl/issues/163.

nalimilan commented on test/broadcast.jl:
```diff
@@ -111,5 +94,52 @@ module TestBroadcast
+    X[1] = Nullable()
+    A[1]= 0
+    # FIXME: element-wise syntax should give the same result
+    # R = get.(X, 0)
```
Cf. https://github.com/JuliaLang/julia/issues/19313.

nalimilan commented:
@davidagold Where do you think `lift` should live? In NullableOps?

davidagold commented:
Preferably NullableArrays. Or am I wrong for thinking 'lift' and lifted ops
are sufficiently different strategies that they ought to live in different
packages?

On Sunday, November 13, 2016, Milan Bouchet-Valat notifications@github.com
wrote:

> @davidagold https://github.com/davidagold Where do you think lift
> should live? In NullableOps?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/JuliaStats/NullableArrays.jl/pull/166#issuecomment-260188991,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ALUCm4fyI9IQ1hlGuon_zLjcVTdMHRJPks5q9xxvgaJpZM4Kv-IA
> .

nalimilan commented:
No, that's fine with me, I just wanted to be sure that was OK for you.

nalimilan commented:
https://github.com/JuliaLang/julia/issues/19313 is kind of annoying: in some cases, the `f.(x)` syntax doesn't work for two-argument functions.

There's another related issue which is even more annoying: loop fusion prevents `lift` from detecting special (blacklisted) functions. So e.g. `identity.(isnull.(x))` returns `NULL` instead of `false` for missing entries.

These two issues mostly affect corner cases, so this change is probably worth it anyway, but in the longer term I'm not sure what to do. The `f.(x)` syntax could be improved in Julia to be more flexible, i.e. calling a method with the function and input types as arguments, which could return a transformed function (lifted, in our case). Then loop fusion would combine these functions instead of the original ones. But I don't see this being accepted in Base without a strong use case, i.e. more or less official support for the lifting framework; we may as well merge https://github.com/JuliaLang/julia/pull/18758 at the same time.

I'm starting to think (again) that all function calls should lift by default; that would be a much more consistent option.

johnmyleswhite commented:
I'm not sure we should move forward with this just yet. It might be the right decision (and I'm inclined to think it is), but it might be easier to start with a less ambitious approach in which people have to explicitly call special functions like `isnull`. People love to complain about the burdens of those kinds of things, but I worry that this approach will be too magical and will actually prevent people from learning how to handle non-standard semantics for lifting.

nalimilan commented:
So that means lifting by default, without any blacklist?

bramtayl commented:
This kind of thing could be integrated pretty easily in ChainMap. Once a lift function is available, I could make

``` julia
@chain begin
    x
    @over is.null
    @lift_over identity
end
```

go to

`lift(broadcast, identity, broadcast(is.null, x) )`

Edit: works now on master

nalimilan committed:
Rewrite broadcast() using Base implementation and lifting semantics

Remove the custom implementation of broadcast(), and just call
the base method on the lifted method. For now, standard lifting
semantics are always used, even for logical operators and isnull/get.

nalimilan commented:
I've updated the branch to always lift. It now uses a private `broadcast_lift` function, since for now we don't want people to override the lifting behaviour.

~~Surprisingly, new implementation appears to be just as fast as the current one. Since it passes the tests on 0.5 and 0.6, I think we can seriously consider merging it. It's not yet clear how to get non-standard lifting to work, but fixing standard cases sounds like the highest priority to me.~~
EDIT: I got the benchmarks wrong. There's still an allocation problem which kills performance. I'll investigate.

nalimilan committed:
Use generated function to avoid allocation

This only covers one- and two-argument cases.

nalimilan committed:
Fix performance without generated functions

nalimilan commented:
Really good news: with the latest changes, I get a very similar performance to master, even slightly better, for one- and two-argument cases. Three-argument and beyond is still terribly slow; I can add a special case for 3 if needed, waiting for a more general solution.

Example on a simple benchmark:
```julia
using NullableArrays
X = NullableArray(1:100000);
Y = NullableArray(1:100000);
using BenchmarkTools

# With this branch:
julia> @benchmark broadcast!(+, X, Y)
BenchmarkTools.Trial: 
  memory estimate:  176.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     107.072 μs (0.00% GC)
  median time:      108.384 μs (0.00% GC)
  mean time:        114.363 μs (0.00% GC)
  maximum time:     360.646 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark broadcast!(+, X, X, Y)
BenchmarkTools.Trial: 
  memory estimate:  224.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     222.965 μs (0.00% GC)
  median time:      228.248 μs (0.00% GC)
  mean time:        246.007 μs (0.00% GC)
  maximum time:     1.621 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

# With master:
julia> @benchmark broadcast!(+, X, Y)
BenchmarkTools.Trial: 
  memory estimate:  224.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     117.858 μs (0.00% GC)
  median time:      151.116 μs (0.00% GC)
  mean time:        197.762 μs (0.00% GC)
  maximum time:     1.576 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark broadcast!(+, X, X, Y)
BenchmarkTools.Trial: 
  memory estimate:  272.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     243.590 μs (0.00% GC)
  median time:      251.746 μs (0.00% GC)
  mean time:        265.030 μs (0.00% GC)
  maximum time:     1.100 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

```

More generally, running `perf/broadcast.jl` shows a slight advantage to this PR, except in a handful of cases where the difference isn't that large:
https://gist.github.com/nalimilan/57ea0640ffb8fc4a8424b558efe3ff4d

Comments?

nalimilan committed:
Improve efficiency of null checking for generic case

In my benchmarks, this approach is just as fast as manually repeating
isnull(x) | isnull(y)...

nalimilan committed:
Fix call to null_safe_op() for generic case

nalimilan committed:
Fix remaining allocations in generic method

Sacha0 commented on src/broadcast.jl:
```diff
@@ -8,199 +9,143 @@ else
+    ftype(f, A) = typeof(f)
+    ftype(f, A...) = typeof(a -> f(a...))
+    ftype(T::DataType, A) = Type{T}
+    ftype(T::DataType, A...) = Type{T}
```
The changes in https://github.com/JuliaLang/julia/pull/19421 might be relevant here. Best!

Sacha0 commented on src/broadcast.jl:
```diff
@@ -8,199 +9,143 @@ else
     end
+    ziptype(A) = Tuple{eltype(A)}
+    ziptype(A, B) = Zip2{Tuple{eltype(A)}, Tuple{eltype(B)}}
+    @inline ziptype(A, B, C, D...) = Zip{Tuple{eltype(A)}, ziptype(B, C, D...)}
```
Likewise here, the changes in https://github.com/JuliaLang/julia/pull/19421 might be relevant. Best!

nalimilan committed:
Add non-splatting methods up to 7 arguments

nalimilan committed:
Also port map() to lift()

This introduces a small behavior change for consistency with the generic
map(): mapping over an empty array returns a NullableArray{Any} when the
function does not exist, rather than a NullableArray{Union{}}.

nalimilan committed:
Fix broadcast() tests

nalimilan committed:
Small naming improvement

nalimilan commented:
I've added the corresponding changes for `map`. Since the current implementation directly accesses the `values` and `isnull` fields, it is about twice faster as the new one. It's a bit annoying, but I'd say we should do it anyway, as we don't really have the manpower to maintain this while improving at the same time the API. With some luck, the compiler will also improve.

Calling the generic `map` implementation required changing the semantics a bit when an array is empty and no function matches the element type: we used to choose a `Union{}` eltype, while the generic implementation returns `Any`. I don't think it matters a lot anyway, but better be consistent.

EDIT: failure on 0.6 is already on master, and fixed by https://github.com/JuliaStats/NullableArrays.jl/pull/169.

davidagold commented on REQUIRE:
```diff
@@ -1,3 +1,4 @@
 julia 0.4
```
Should we delete this line?

davidagold commented on src/broadcast.jl:
```diff
@@ -8,199 +9,92 @@ else
-    """
-        broadcast!(f, B::NullableArray, As::NullableArray...; lift::Bool=false)
+invoke_broadcast!{F, N}(f::F, dest, As::Vararg{NullableArray, N}) =
+    invoke(broadcast!, Tuple{Function, AbstractArray, Vararg{AbstractArray, N}}, f, dest, As...)
```
Should `Function` be `F`? Probably won't make a difference in most cases.

nalimilan replied:
Probably, I guess it's more correct if e.g. `f` is a type and there's a specific method for them.

davidagold commented on src/broadcast.jl:
```diff
@@ -8,199 +9,92 @@ else
-                                       _to_shape(broadcast_indices(Xs...))),
-                      Xs...; lift=lift)
+function Base.broadcast!{F, N}(f::F, dest::NullableArray, As::Vararg{NullableArray, N})
+    # These definitions are needed to avoid allocation due to splatting
```
Have you verified that allocations are avoided for `f2` of more than two arguments? Also, did you try `@inline` decorations for `f2` to avoid the splatting penalty?

nalimilan replied:
There are no allocations on Julia master, though beyond 2 arguments performance is lower. I think I've tried everything I could (given my skills at least), without success (except a generated function, but that's probably overkill for one line). Since there are several issues open about varargs being slow, my hope is that it's going to improve at some point.

davidagold commented on src/lift.jl:
```diff
@@ -0,0 +1,61 @@
+    if null_safe_op(f, eltype(x))
+        return @compat Nullable(f(unsafe_get(x)), !isnull(x))
+    else
+        U = Core.Inference.return_type(f, Tuple{eltype(x)})
```
We should probably check if `U` is leaf, yes? If so, we should decide what to do when `U` is abstract. Return `Union{}` as in 16961?

nalimilan replied:
Makes sense.

davidagold commented on src/lift.jl:
```diff
@@ -0,0 +1,61 @@
+        if hasnulls(xs...)
+            return Nullable{U}()
+        else
+            return Nullable(f(_unsafe_get(xs...)...))
```
Out of curiosity, why the recursive approach as opposed to `map(unsafe_get, xs)...` ?

nalimilan replied:
In my tests it avoided allocations, which isn't the case with `map`.

davidagold commented on src/broadcast.jl:
```diff
@@ -8,199 +9,92 @@ else
-    #   broadcast!(f, Z::NullableArrays.NullableArray, Xs::NullableArrays.NullableArray...)
-    @inline Base.broadcast!(f, Z::NullableArray; lift=false) =
-        broadcast!(f, Z, Z; lift=lift)
+Call `broadcast` with nullable lifting semantics and return a `NullableArray`.
```
We should revise this line in light of the of the behavior defined in line 62.

nalimilan replied:
Good catch. Actually, since in this PR for now `lift` isn't supposed to be overloaded (so that lifting is always done), the description is correct. I've removed line 62. I hope we can improve this later in another PR.

davidagold commented on src/map.jl:
```diff
@@ -1,248 +1,67 @@
+"""
+    map(f, As::NullableArray...)
+
+Call `map` with nullable lifting semantics and return a `NullableArray`.
```
Similarly should revise this description.

davidagold commented on src/map.jl:
```diff
@@ -1,248 +1,67 @@
+using Base: collect_similar, Generator
+
+invoke_map!{F, N}(f::F, dest, As::Vararg{NullableArray, N}) =
+    invoke(map!, Tuple{Function, AbstractArray, Vararg{AbstractArray, N}}, f, dest, As...)
```
Similarly, should we have `Function` -> `F`?

davidagold commented on src/operators.jl:
```diff
@@ -37,8 +36,7 @@ Types declared as safe can benefit from higher performance for operations on nul
 """
-null_safe_op(f::Any, ::Type) = false
-null_safe_op(f::Any, ::Type, ::Type) = false
+null_safe_op(f::Any, ::Type...) = false
```
I'm confused. Isn't this in Base?

nalimilan replied:
We still carry a copy of `null_safe_op` because it doesn't exist on 0.4, and even on 0.5 it only comes with definitions for `isless` and `isequal`. But indeed we should merge them at some point (I'll do that in another PR).

nalimilan committed:
Review comments

nalimilan committed:
Use generated function to avoid allocations on Julia 0.5

The 1- and 2-argument versions are slower on 0.5 due to allocations,
but faster (especially the latter) on 0.6.

nalimilan commented:
I've just added a commit which uses a generated function for the general `lift` method to get rid of allocations on Julia 0.5 (interestingly, the 1- and 2-argument versions do allocate there, contrary to 0.6). Unfortunately, it's still quite noticeably slower than on 0.6 (about 6x for the 1-argument version), but still "reasonable".

So I think this is now good to go if we agree on the strategy.

nalimilan committed:
Simplify code and improve performance on Julia 0.5, use only generated function

Performance is now the same on Julia 0.5 and 0.6, and the 1- and 2-argument
specialization give the same result.

nalimilan commented:
Good news! By replacing varargs with a tuple in `lift`, I've been able to get the best performance without specialized 1- and 2-argument functions, on both Julia 0.5 and 0.6. So now only the generated function remains, and `lift` give really clean native code:
```julia
julia> @code_native NullableArrays.lift(+, (Nullable(1),Nullable(2)))
	.text
Filename: lift.jl
	pushq	%rbp
	movq	%rsp, %rbp
Source line: 21
	movq	24(%rsi), %rax
	addq	8(%rsi), %rax
Source line: 22
	movb	(%rsi), %cl
	andb	16(%rsi), %cl
Source line: 23
	andb	$1, %cl
Source line: 17
	movb	%cl, (%rdi)
	movq	%rax, 8(%rdi)
	movq	%rdi, %rax
	popq	%rbp
	retq
	nop
```

The three-argument `map` still allocates a lot currently, but this is easy to fix: https://github.com/JuliaLang/julia/pull/19522. Other than that, `map` is generally about twice slower than `broadcast`, but I'm not yet sure why. It's most likely separate from this PR anyway since both functions use the same approach.

More comments before merging?

nalimilan committed:
Remove Julia 0.4 lines

Rebasing mistake probably.

p-zubieta commented:
@nalimilan Just a heads-up as I found a way of simplifying the promotion mechanism for broadcast. You can check it out here https://github.com/JuliaLang/julia/pull/19421. So you might want to wait until that lands.

nalimilan commented:
Yes, thanks, @Sacha0 posted the link above too. I'll update this PR once the other is merged.

nalimilan committed:
Update eltype computation code to match Base

These have recently been simplified. map() in Base does not use
the same path as broadcast(), but doing so here is much simpler
and results should be equivalent for consistency anyway.

nalimilan committed:
Fix 0.6 deprecations

nalimilan committed:
Fix deprecation warnings

nalimilan commented:
I've pushed a few more commits to reflect recent changes in Base. I've also included 0.6 fixes from my other PRs, unfortunately, this depends on https://github.com/JuliaLang/Compat.jl/pull/306. Finally, I've tweaked our `nullable_broadcast_eltype` to follow https://github.com/JuliaLang/julia/pull/20049, since we use the same code for `broadcast` and `map` (could have to change if that PR isn't merged).

So I think it's good to go after the new Compat release. Any more comments?

davidagold commented:
None from me, I don't think. I'm very excited about this.

nalimilan committed:
Require Compat 0.13.0

nalimilan committed:
Fix computing eltype

nalimilan commented:
Should be good to go now, the 0.6 failure is in operators.jl.

p-zubieta commented on src/broadcast.jl:
```diff
@@ -25,7 +25,7 @@ if VERSION < v"0.6.0-dev" # Old approach needed for inference to work
 
     nullable_broadcast_eltype(f, As...) =
-        _default_eltype(Base.Generator{ziptype(As...), ftype(f, As...)})
+        eltype(_default_eltype(Base.Generator{ziptype(As...), ftype(f, As...)}))
```
I'm curious why is this needed?

nalimilan replied:
What is "this"?

p-zubieta replied:
Sorry, the `eltype` you last added.

nalimilan replied:
It's useful because `NullableArray{T}` is parameterized on the eltype of `Nullable` rather than on the `Nullable{T}` itself.

p-zubieta replied:
But, wouldn't that warrants that you don't get a `Nullable{T}` here?

Given that `eltype(NullableArray([1], [true])) == Int`, that implies `_default_eltype(+, NullableArray([1], [true]), NullableArray([1], [true])) == Int`, right?

nalimilan replied:
No, `eltype(NullableArray([1], [true])) == Nullable{Int}`.

I don't want `T = Nullable{Int}` here since it's used to build a new array via `NullableArray{T}(...)`. `NullableArray{Nullable{Int}}` would have a `Nullable{Nullable{Int}}` eltype.

p-zubieta replied:
The problem with this is that `eltype(_default_eltype(*, ["a"], ["b"])` will give `Char`, so I guess you'd need a custom `eltype` here.

nalimilan replied:
But that method is only called on `NullableArrays`, so we're certain the element types are `Nullable` (in your case, `Nullable{String}`).

p-zubieta replied:
Oh, now I see, this is `nullable_broadcast_eltype` as opposed to `broadcast_eltype`. Sorry for the noise then.

nalimilan commented:
Will squash and merge tomorrow unless somebody objects or beats me to it.

tkelman commented on test/map.jl:
```diff
@@ -83,79 +65,59 @@ module TestMap
-    if VERSION >= v"0.5.0-dev+3294"
-        @test isequal(Z1, NullableArray(Union{}[])) # no type inference
+    @test isempty(Z1)
+    if VERSION >= v"0.6.0-dev"
```
please be specific with VERSION cutoffs like this, will this really work for all 0.6-dev versions?

nalimilan replied:
Actually, looks like I fixed this in https://github.com/JuliaStats/NullableArrays.jl/pull/177/files. Will remove the `if`.

(FWIW, at the time I added these conditions, the issue was that inference failed on 0.5.0, and I had no idea which commit(s) fixed it except for bisecting it...)

nalimilan replied:
See https://github.com/JuliaStats/NullableArrays.jl/pull/183.
